edge-modules/api-proxy-module/src/monitors/token_client.rs

```diff
@@ -0,0 +1,142 @@
+        .context(format!("Missing env var {}", "IOTEDGE_WORKLOADURI"))?;
+
+    TokenClient::new(
+        device_id,
+        module_id,
+        generation_id,
+        iothub_hostname,
+        &workload_url,
+        validity_hours,
+    )
+}
+
+pub const IOTHUB_ENCODE_SET: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS // C0 control
+    .add(b' ')
+    .add(b'"')
+    .add(b'<')
+    .add(b'>')
+    .add(b'`') // fragment
+    .add(b'#')
+    .add(b'?')
+    .add(b'{')
+    .add(b'}') // path
+    .add(b'=');
+
+pub struct TokenClient {
+    device_id: String,
+    module_id: String,
+    generation_id: String,
+    iothub_hostname: String,
+    work_load_api_client: edgelet_client::WorkloadClient,
+    token_expiration_date: Option<DateTime<Utc>>,
+    validity_duration: Duration,
+}
+
+impl TokenClient {
+    pub fn new(
+        device_id: String,
+        module_id: String,
+        generation_id: String,
+        iothub_hostname: String,
+        workload_url: &str,
+        validity_duration: Duration,
+    ) -> Result<Self, Error> {
+        //Check duration more than margin
+
+        //Create expiry date in the past so cert has to be rotated now.
+        let token_expiration_date = None;
+
+        let work_load_api_client =
+            edgelet_client::workload(&workload_url).context("Could not get workload client")?;
+
+        Ok(TokenClient {
+            device_id,
+            module_id,
+            generation_id,
+            iothub_hostname,
+            work_load_api_client,
+            token_expiration_date,
+            validity_duration,
+        })
+    }
+
+    pub async fn get_new_sas_token(
+        &mut self,
+        current_date: DateTime<Utc>,
+    ) -> Result<Option<String>, Error> {
+        //If there is no previous token request a new one right away
+        if let Some(expiration_date) = self.token_expiration_date {
+            //If Token is not expired, sleep until it is.
+            use std::convert::TryInto;
+            if current_date < expiration_date {
+                let delay_seconds: u64 = expiration_date
+                    .signed_duration_since(current_date)
+                    .num_seconds()
+                    .try_into()
+                    .unwrap_or(0);
+
+                let delay_seconds = tokio::time::Duration::from_secs(delay_seconds);
+                time::delay_for(delay_seconds).await;
+            } else {
+                error!("Token has expired before API proxy was able to request a new one. Requesting a new SAS token now");
+            }
+        }
+
+        info!("Generating new token");
+        // Add duration time and remove the margin.
+        let new_expiration_date = Utc::now()
+            .checked_add_signed(self.validity_duration)
+            .context("Could not compute new expiration date for certificate")?
+            .checked_sub_signed(Duration::seconds(TOKEN_EXPIRY_SECONDS_MARGIN))
```

Comment:
Why are you subtracting the margin secs here?

Reply:
This is the same as above, let's chat about it